updater.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
+
+import os
+import subprocess
+
+service_name = 'redeem'
+
+
+def switch_to_src_directory():
+  # assemble the path to the source directory
+  this_dir = os.path.dirname(os.path.realpath(os.path.abspath(__file__)))
+  source_location = os.path.normpath(os.path.join(this_dir, "."))
+  os.chdir(source_location)
+
+
+def is_service_running(name):
+  with open(os.devnull, 'wb') as hide_output:
+    exit_code = subprocess.Popen(['sudo', 'service', name, 'status'],
+                                 stdout=hide_output,
+                                 stderr=hide_output).wait()
+    return exit_code == 0
+
+
+def start_service(name):
+  print("Start the {} service".format(name))
+  with open(os.devnull, 'wb') as hide_output:
+    exit_code = subprocess.Popen(['sudo', 'service', name, 'start'],
+                                 stdout=hide_output,
+                                 stderr=hide_output).wait()
+    return exit_code == 0
+
+
+def stop_service(name):
+  print("Stop the {} service".format(name))
+  with open(os.devnull, 'wb') as hide_output:
+    exit_code = subprocess.Popen(['sudo', 'service', name, 'stop'],
+                                 stdout=hide_output,
+                                 stderr=hide_output).wait()
+    return exit_code == 0
+
+
+def perform_git_update():
+  print("Do git update here")
+  with open(os.devnull, 'wb') as hide_output:
+    exit_code = subprocess.Popen(['git', 'pull'], stdout=hide_output, stderr=hide_output).wait()
+    return exit_code == 0
+
+
+def reinstall():
+  print("Run setup")
+  with open(os.devnull, 'wb') as hide_output:
+    exit_code = subprocess.Popen(['python', 'setup.py', 'develop'],
+                                 stdout=hide_output,
+                                 stderr=hide_output).wait()
+    return exit_code == 0
+
+
+def perform_update():
+  switch_to_src_directory()
+  service_was_running = is_service_running(service_name)
+  if service_was_running:
+    stop_service(service_name)
+  perform_git_update()
+  reinstall()
+  if service_was_running:
+    start_service(service_name)
+  print("Update complete")
+
+
+if __name__ == '__main__':
+  perform_update()
```

Comment:
Isn't all this literally the same as...
```
#!/bin/bash

SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
service redeem status >> /dev/null
REDEEM_RUNNING=$?

set -e

cd $SCRIPT_DIR
if [ "$REDEEM_RUNNING" -eq "0"]; then service redeem stop; fi
git pull
python setup.py develop
if [ "$REDEEM_RUNNING" -eq "0"]; then service redeem start; fi
echo "Update complete"
```

Reply:
No, it is not "literally the same" --
First, your bash script does not take into consideration the python virtual environment.
Second, although they are functionally similar, you are switching ecosystems.  We could have written this functionality in Ruby, JavaScript, Go, or any of a number of other languages. However, since this is a PYTHON project, it is appropriate to remain in the python ecosystem.  Instead of embracing the python infrastructure, you seem to want to do things the C++ way and force the python code to conform to that methodology.

Reply:
> No, it is not "literally the same" --
> First, your bash script does not take into consideration the python virtual environment.

It does in exactly the same way the python script does. If you invoke either while inside a virtualenv, they will execute inside the virtualenv.

> Second, although they are functionally similar, you are switching ecosystems. We could have written this functionality in Ruby, JavaScript, Go, or any of a number of other languages. However, since this is a PYTHON project, it is appropriate to remain in the python ecosystem. Instead of embracing the python infrastructure, you seem to want to do things the C++ way and force the python code to conform to that methodology.

Writing bash scripts to run command-line tools on a Linux system is "the C++ way" ? That's a new one.

You're invoking subprocesses with executables on the $PATH, suppressing their output, and doing basic error checking (and then ignoring the errors in `perform_update()`). Python isn't necessarily bad at that, but that's exactly what bash was designed to do. I advocate choosing the best tool for the job.

Reply:
"If you invoke either while inside a virtualenv" ... But the python version will also execute in its proper virtual environment without first doing any explicit action to enter that environment.

By "the C++ way", I mean that using the build system commonly associated with the particular language. Each of the major programming languages have developed their own set of routines which fundamentally accomplish the same thing. However, each also has features which are particularly oriented to their own language. Using `Make` and `Bash` scripts are more closely associated with C++. Python, Ruby, Java, JavaScript each have scripting written in their own language.

Reply:
> Using Make and Bash scripts are more closely associated with C++. Python, Ruby, Java, JavaScript each have scripting written in their own language.
Uh, no. I would argue that Make and Bash scripts are general deployment systems common to Linux. I've seen applications written in python other than what we have in the `intelligent-agent` repository provide `Makefile`s in a consistent way, generally, when, like Redeem, the code is mixing Python with C or C++ bindings.

Pip, npm, maven etc. are more often meant to be used as _distribution_ systems, not build systems. In that way they're language-specific tooling that have more in common with apt, yum, AppImage or others. `Make` has more in common with `ant` than it does with `maven`. It's not there to do dependency management, but specifically to build code.

It's a staple of any Linux/Unix-specific application, and by using it we are making use of a (admittedly rather informal) standard that is common on that platform.

Reply:
Yes, as you mention, those are distribution tools. But `make`, `ant`, `setup`, etc. are build tools. And since we are not working in a "distribution" mode (installing executables built on another system), we need to perform both the source to executable transformation and also place the resulting executable in an operating environment. Just because you can use the build tools from one environment in some other environment, it does not mean that you should do so. But, since your interest is only in building a particular system, working around the ecosystem to shoehorn it into your methodology works for you. However, I would suggest that if you insist on avoiding working with the language's ecosystem, you drop the use of Python altogether. There are better languages (including C++) for doing what you are doing. They generate more efficient code, etc. Pick a system. Then work WITH it, don't try to force it to behave like some other system.

Reply:
I'm seemingly not getting the point I wanted across, so let me put this a different way.
My objections to this code as it stands are that actual operations are buried in boilerplate and the error handling is broken. If I refactor it to fix those, it ends up looking like this:
```
#!/usr/bin/env python

import os
import subprocess

service_name = 'redeem'

def run_process(cmdline):
  with open(os.devnull, 'wb') as hide_output:
    exit_code = subprocess.Popen(cmdline.split(" "),
                                 stdout=hide_output,
                                 stderr=hide_output).wait()
    if exit_code != 0:
      raise RuntimeError("command failed: " + cmdline)

def switch_to_src_directory():
  # assemble the path to the source directory
  this_dir = os.path.dirname(os.path.realpath(os.path.abspath(__file__)))
  source_location = os.path.normpath(os.path.join(this_dir, "."))
  os.chdir(source_location)


def is_service_running(name):
  try:
    run_process("sudo service {} status".format(service_name))
    return true
  except:
    return false


def start_service(name):
  print("Start the {} service".format(name))
  run_process("sudo service {} start".format(service_name))


def stop_service(name):
  print("Stop the {} service".format(name))
  run_process("sudo service {} stop".format(service_name))


def perform_git_update():
  print("Do git update here")
  run_process("git pull")


def reinstall():
  print("Run setup")
  run_process("python setup.py develop")


def perform_update():
  switch_to_src_directory()
  service_was_running = is_service_running(service_name)
  if service_was_running:
    stop_service(service_name)
  perform_git_update()
  reinstall()
  if service_was_running:
    start_service(service_name)
  print("Update complete")


if __name__ == '__main__':
  perform_update()
```

... but now it becomes clear that all the two-line functions are actually just obscuring what the script does instead of abstracting anything, so it refactors to the much shorter...

```
#!/usr/bin/env python

import os
import subprocess

service_name = 'redeem'

def run_process(cmdline):
  with open(os.devnull, 'wb') as hide_output:
    exit_code = subprocess.Popen(cmdline.split(" "),
                                 stdout=hide_output,
                                 stderr=hide_output).wait()
    if exit_code != 0:
      raise RuntimeError("command failed: " + cmdline)

def perform_update():
  # assemble the path to the source directory
  this_dir = os.path.dirname(os.path.realpath(os.path.abspath(__file__)))
  source_location = os.path.normpath(os.path.join(this_dir, "."))
  os.chdir(source_location)
  
  redeem_running = false
  
  try:
    run_process("sudo service {} status".format(service_name))
    redeem_running = true
  except:
    pass

  if redeem_running:
    run_process("sudo service {} stop".format(service_name))

  run_process("git pull")
  run_process("python setup.py develop")

  if redeem_running:
    run_process("sudo service {} start".format(service_name))

  print("Update complete")


if __name__ == '__main__':
  perform_update()
```

... but now it becomes very clear that there's nothing pythonic about what this script is doing. We're invoking...
- `sudo`, which is an ELF binary
- `service`, which is a shell script
- `git`, which is another ELF binary
- `python`, which is another ELF binary

... and it's actually worse than that because we've broken sudo by piping it to /dev/null, so it can't possibly prompt a non-root user for credentials anyway.

At this point I wrote a bash script because it's more succinct, more transparent, and it's better at invoking ELF binaries and shell scripts than python will ever be.

I don't understand why you believe using the standard Linux tools to do Linux systems management tasks is "working around the ecosystem to shoehorn it into your methodology." We use python when it makes our system easier to understand and modify. We use C++ when performance considerations are more critical than ease of understanding and modification. We use other tools when they are the most appropriate tools for the task at hand.

Put another way, I can easily see that putting this functionality into a shell script is more comprehensible and less error-prone. I cannot see what putting this functionality in a python script gains us.